outlines/generate/generator.py

```diff
@@ -0,0 +1,362 @@
+    ) -> Union[str, List[str]]:
+        sequence_generator = self.stream(prompt, kv_cache, rng)
+        tokens = [token for token in sequence_generator]
+        sequences = [
+            self.format_sequence("".join(sequence)) for sequence in list(zip(*tokens))
+        ]
+        return sequences if len(sequences) > 1 else sequences[0]
+
+    def stream(
+        self,
+        prompt: str,
+        kv_cache: Optional[torch.tensor] = None,
+        rng: Optional[torch.Generator] = None,
+    ) -> Iterator[Union[List[str], str]]:
+        if rng is None:
+            rng = torch.Generator(device=self.device)
+            rng.seed()
+
+        init_state = self.init_generation_state(prompt, kv_cache)
+
+        token_ids = init_state[1]
+        num_sequences = token_ids.shape[0]
+
+        init_fsm_states = [FSMState(0) for _ in range(num_sequences)]
+
+        states = sequence_generator(
+            self.generate_token, self.fsm, init_state, init_fsm_states, rng
+        )
+
+        def token_generator() -> Iterator[Union[List[str], str]]:
+            while True:
+                try:
+                    sequence = next(states)
+                except StopIteration:
+                    return
+
+                next_token_ids = sequence.token_ids[:, -1]
+                next_tokens = self.tokenizer.decode(next_token_ids)
+
+                yield next_tokens
+
+        return token_generator()
+
+
+def sequence_generator(
+    token_generator: Callable,
+    fsm: "FSM",
+    init_state: Tuple,
+    fsm_states: List[FSMState],
+    rng: torch.Generator,
+) -> Iterator[GenerationState]:
+    """Generates sequences of tokens.
+
+    Parameters
+    ----------
+    token_generator
+        A callable that generate a new token given the current generation state
+        and logits biases.
+    fsm
+        The finite-state machine that drives the text generation.
+    init_state
+        The initial generation state for the batches.
+    fsm_states
+        The initial states of the finite-state machine for each sequence in the batch.
+
+    Yields
+    ------
+    A new sequence.
+
+    """
+    token_ids, attention_masks, kv_cache = init_state
+    while True:
+        logits_masks = get_next_instruction(fsm, fsm_states)
+
+        next_token_ids, kv_cache, logits = token_generator(
+            token_ids,
+            attention_masks,
+            kv_cache,
+            rng=rng,
+            logits_masks=logits_masks,
+        )
+
+        token_ids = update_token_ids(token_ids, next_token_ids)
+        attention_masks = expand_attention_masks(attention_masks)
+
+        fsm_states = get_next_fsm_states(fsm, fsm_states, next_token_ids)
+        is_finished = is_generation_finished(fsm, fsm_states)
+
+        if is_finished:
+            yield GenerationState(token_ids, kv_cache, logits)
+            return
+
+        yield GenerationState(token_ids, kv_cache, logits)
+
+
+@torch.inference_mode
```

Comment:
Should this instead be on the closure `generate` created by this constructor function?  My understanding is that `inference_mode` creates a context manager that's in effect for the duration of the call to `token_generator`, but we want one that applies to the function constructed by `token_generator` (i.e. `generate`).  Perhaps I'm forgetting something that propagates that context manager and/or its effect?